src/Data/Avro/Encoding/ToAvro.hs

```diff
@@ -0,0 +1,226 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleInstances         #-}
+{-# LANGUAGE OverloadedStrings         #-}
+{-# LANGUAGE RankNTypes                #-}
+{-# LANGUAGE ScopedTypeVariables       #-}
+{-# LANGUAGE TypeApplications          #-}
+{-# LANGUAGE TypeFamilies              #-}
+
+
+module Data.Avro.Encoding.ToAvro
+where
+
+import           Control.Monad.Identity       (Identity (..))
+import qualified Data.Array                   as Ar
+import           Data.Avro.Internal.EncodeRaw
+import           Data.Avro.Internal.Time
+import           Data.Avro.Schema.Decimal     as D
+import           Data.Avro.Schema.Schema      as S
+import qualified Data.Binary.IEEE754          as IEEE
+import qualified Data.ByteString              as B
+import           Data.ByteString.Builder
+import           Data.ByteString.Lazy         as BL
+import qualified Data.Foldable                as F
+import           Data.HashMap.Strict          (HashMap)
+import qualified Data.HashMap.Strict          as HashMap
+import           Data.Int
+import           Data.Ix                      (Ix)
+import           Data.List                    as DL
+import qualified Data.Map.Strict              as Map
+import           Data.Maybe                   (fromJust)
+import           Data.Text                    (Text)
+import qualified Data.Text                    as T
+import qualified Data.Text.Encoding           as T
+import qualified Data.Text.Lazy               as TL
+import qualified Data.Text.Lazy.Encoding      as TL
+import qualified Data.Time                    as Time
+import qualified Data.UUID                    as UUID
+import qualified Data.Vector                  as V
+import qualified Data.Vector.Unboxed          as U
+import           Data.Word
+import           GHC.TypeLits
+
+newtype Encoder = Encoder { runEncoder :: Schema -> Builder }
+
+(.=) :: forall a. ToAvro a => Text -> a -> (Text, Encoder)
+(.=) fieldName fieldValue = (fieldName, Encoder (flip toAvro fieldValue))
+
+record :: Schema -> [(Text, Encoder)] -> Builder
+record (S.Record _ _ _ _ fs) vs =
+  foldMap (mapField provided) fs
+  where
+    provided :: HashMap Text Encoder
+    provided = HashMap.fromList vs
+
+    providedNames = fst <$> vs
+
+    failField :: S.Field -> Builder
+    failField fld = error $ "Field '" <> show (S.fldName fld) <> "' is missing from the provided list of fields: " <> show providedNames
+
+    mapField :: HashMap Text Encoder -> S.Field -> Builder
+    mapField env fld =
+      maybe (failField fld) (flip runEncoder (S.fldType fld)) (HashMap.lookup (S.fldName fld) env)
+
+-- | Describes how to encode Haskell data types into Avro bytes
+class ToAvro a where
+  toAvro :: Schema -> a -> Builder
+
+instance ToAvro Int where
+  toAvro (S.Long _) i = encodeRaw @Int64 (fromIntegral i)
+  toAvro (S.Int _) i  = encodeRaw @Int32 (fromIntegral i)
+  toAvro s _          = error ("Unable to encode Int as: " <> show s)
```

Comment:
Shouldn't you be able to encode an `Int` into a schema that expects a `Double`?  Or is that somewhere else...

Reply:
Hmm, you are right! I'll see through this and potentially other missing cases. Thanks!